assembly/src/parsers/io_ops.rs

```diff
@@ -17,16 +17,36 @@ const HEX_CHUNK_SIZE: usize = 16;
 // INSTRUCTION PARSERS
 // ================================================================================================
 
-/// Returns `PushConstants` instruction node.
+/// Returns one of the `Push` instruction nodes.
 ///
 /// # Errors
 /// Returns an error if the instruction token has invalid values or inappropriate number of
 /// values.
 pub fn parse_push(op: &Token) -> Result<Node, ParsingError> {
-    validate_operation!(op, "push", 1..MAX_PUSH_INPUTS);
-
-    let constants = parse_constants(op)?;
-    Ok(Instruction(PushConstants(constants)))
+    debug_assert_eq!(op.parts()[0], "push");
+    match op.num_parts() {
+        0 => unreachable!(),
+        1 => Err(ParsingError::missing_param(op)),
+        2 => {
+            let param_str = op.parts()[1];
+            if let Some(param_str) = param_str.strip_prefix("0x") {
+                // if we have only one parameter
+                if param_str.len() <= HEX_CHUNK_SIZE {
+                    // check parameter validity
+                    if param_str.len() % 2 != 0 {
+                        return Err(ParsingError::invalid_param(op, 1));
+                    }
+                    evaluate_type_one(op, parse_hex_param(op, 1, param_str)?)
+                } else {
+                    evaluate_type_many(op, parse_hex_params(op, param_str)?)
+                }
+            } else {
+                evaluate_type_one(op, parse_decimal_param(op, 1, param_str)?)
+            }
+        }
+        3..=17 => evaluate_type_many(op, parse_constants(op)?),
```

Comment:
I tried to do this, but apparently I can't use expressions as an interval boundary. I can create a local constant value, but it looks like a little overkill for its purpose

Reply:
It is usually good to have such constants because if we ever change `MAX_PUSH_CONSTANTS` to something other than 16, there will be fewer places where we'd need to make changes.